pkg/stanza/fileconsumer/internal/archive/archive.go

```diff
@@ -0,0 +1,187 @@
+	// To minimize disk access, we first access the index, then review unmatched files and update the metadata, if found.
+	// We exit if all fingerprints are matched.
+
+	// Track number of matched fingerprints so we can exit if all matched.
+	var numMatched int
+
+	// Determine the index for reading archive, starting from the most recent and moving towards the oldest
+	nextIndex := a.archiveIndex
+	matchedMetadata := make([]*reader.Metadata, len(fps))
+
+	// continue executing the loop until either all records are matched or all archive sets have been processed.
+	for i := 0; i < a.pollsToArchive; i++ {
+		// Update the mostRecentIndex
+		nextIndex = (nextIndex - 1 + a.pollsToArchive) % a.pollsToArchive
+
+		data, err := a.readArchive(ctx, nextIndex) // we load one fileset atmost once per poll
+		if err != nil {
+			a.logger.Error("error while opening archive", zap.Error(err))
+			continue
+		}
+		archiveModified := false
+		for j, fp := range fps {
+			if matchedMetadata[j] != nil {
+				// we've already found a match for this index, continue
+				continue
+			}
+			if md := data.Match(fp, fileset.StartsWith); md != nil {
+				// update the matched metada for the index
+				matchedMetadata[j] = md
+				archiveModified = true
+				numMatched++
+			}
+		}
+		if !archiveModified {
+			continue
+		}
+		// we save one fileset atmost once per poll
+		if err := a.writeArchive(ctx, nextIndex, data); err != nil {
+			a.logger.Error("error while opening archive", zap.Error(err))
+		}
+		// Check if all metadata have been found
+		if numMatched == len(fps) {
+			return matchedMetadata
+		}
+	}
+	return matchedMetadata
+}
+
+func (a *archive) WriteFiles(ctx context.Context, metadata *fileset.Fileset[*reader.Metadata]) {
+	// We make use of a ring buffer, where each set of files is stored under a specific index.
+	// Instead of discarding knownFiles[2], write it to the next index and eventually roll over.
+	// Separate storage keys knownFilesArchive0, knownFilesArchive1, ..., knownFilesArchiveN, roll over back to knownFilesArchive0
+
+	// Archiving:  ┌─────────────────────on-disk archive─────────────────────────┐
+	//             |    ┌───┐     ┌───┐                     ┌──────────────────┐ |
+	// index       | ▶  │ 0 │  ▶  │ 1 │  ▶      ...       ▶ │ polls_to_archive │ |
+	//             | ▲  └───┘     └───┘                     └──────────────────┘ |
+	//             | ▲    ▲                                                ▼     |
+	//             | ▲    │ Roll over overriting older offsets, if any     ◀     |
+	//             └──────│──────────────────────────────────────────────────────┘
+	//                    │
+	//                    │
+	//                    │
+	//                   start
+	//                   index
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(a.archiveIndex); err != nil {
+		a.logger.Error("error faced while encoding the index", zap.Error(err))
```

Comment:
Shorter and clearer?
```suggestion
		a.logger.Error("failed to encode archive index", zap.Error(err))
```